src/order/locally_finite.lean

```diff
@@ -482,6 +482,239 @@ locally_finite_order.of_Icc' (α × β)
 
 end preorder
 
+/-!
+#### `with_top`, `with_bot`
+
+Adding a `⊤` to a locally finite `order_top` keeps it locally finite.
+Adding a `⊥` to a locally finite `order_bot` keeps it locally finite.
+-/
+
+section order_top
+open with_top
+variables (α) [order_top α] [locally_finite_order α]
+
+instance : locally_finite_order (with_top α) :=
+{ finset_Icc := λ a b, match a, b with
+    |      ⊤,      ⊤ := {⊤}
+    |      ⊤, some b := ∅
+    | some a,      ⊤ := cons (⊤ : with_top α) ((Ici a).map embedding.some)
+                              (λ h, let ⟨x, _, hx⟩ := mem_map.1 h in coe_ne_top hx)
+    | some a, some b := (Icc a b).map embedding.some
+    end,
+  finset_Ico := λ a b, match a, b with
+    |      ⊤,      _ := ∅
+    | some a,      ⊤ := (Ici a).map embedding.some
+    | some a, some b := (Ico a b).map embedding.some
+    end,
+  finset_Ioc := λ a b, match a, b with
+    |      ⊤,      _ := ∅
+    | some a,      ⊤ := cons (⊤ : with_top α) ((Ioi a).map embedding.some)
+                              (λ h, let ⟨x, _, hx⟩ := mem_map.1 h in coe_ne_top hx)
+    | some a, some b := (Ioc a b).map embedding.some
+    end,
+  finset_Ioo := λ a b, match a, b with
+    |      ⊤,      _ := ∅
+    | some a,      ⊤ := (Ioi a).map embedding.some
+    | some a, some b := (Ioo a b).map embedding.some
+    end,
+  finset_mem_Icc := begin
+    rintro (_ | a) (_ | b) x,
+    { exact mem_singleton.trans (le_antisymm_iff.trans $ and_comm _ _) },
+    { exact iff_of_false (not_mem_empty _) (λ h, (h.1.trans h.2).not_lt $ some_lt_none _) },
+    { simp only [with_top.le_none, and_true],
+      cases x,
+      { exact iff_of_true (mem_cons_self _ _) le_top },
+      { refine mem_cons.trans ((or_iff_right_of_imp $ λ h, (coe_ne_top h).elim).trans _),
+        rw mem_map,
+        refine ⟨_, λ h, ⟨x, mem_Ici.2 (some_le_some.1 h), rfl⟩⟩,
+        rintro ⟨y, hy, hxy⟩,
+        rw ←hxy,
+        exact some_le_some.2 (mem_Ici.1 hy) } },
+    { refine mem_map.trans ⟨_, λ h, _⟩,
+      { rintro ⟨y, hy, rfl⟩,
+        change some _ ≤ some _ ∧ some _ ≤ some _,
+        rwa [some_le_some, some_le_some, ←mem_Icc] },
+      cases x,
+      { exact (h.2.not_lt $ some_lt_none _).elim },
+      { rw [some_le_some, some_le_some, ←mem_Icc] at h,
+        exact ⟨x, h, rfl⟩ } }
+  end,
```

Comment:
If you use the `coe` change I suggest above, this can be shortened to:
```suggestion
  finset_mem_Icc := λ a b x, match a, b, x with
    | ⊤,       ⊤,       x       := mem_singleton.trans (le_antisymm_iff.trans $ and_comm _ _)
    | ⊤,       (b : α), x       := iff_of_false (not_mem_empty _) (λ h, (h.1.trans h.2).not_lt $ some_lt_none _)
    | (a : α), ⊤,       ⊤       := by simp [with_top.locally_finite_order._match_1]
    | (a : α), ⊤,       (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
    | (a : α), (b : α), ⊤       := by simp [with_top.locally_finite_order._match_1]
    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
  end,
```
although obviously that's more `simp` heavy.

Reply:
The golf is probably worth the heavy simping, however isn't referring to autogenerated names bad practice?

Reply:
I don't think it's bad practice for `match`, where you don't really have any alternative beyond a bunch of private definitions

Reply:
Hmm...
```
    | (a : α), ⊤,       (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
    | (a : α), (b : α), ⊤       := by simp [with_top.locally_finite_order._match_1]
    | (a : α), (b : α), (x : α) := by simp [with_top.locally_finite_order._match_1, coe_eq_coe]
```
doesn't work for me.

Reply:
Even with [this change](https://github.com/leanprover-community/mathlib/pull/10202#discussion_r744246653)? What does it get stuck on?

Reply:
Yup, even with the other change. The precise error is
```
tactic failed, there are unsolved goals
state:
α : Type u_1,
_inst_1 : order_top α,
_inst_2 : locally_finite_order α,
a b x : with_top α,
_match :
  ∀ (_a _a_1 _a_2 : with_top α),
    _a_2 ∈ with_top.locally_finite_order._match_1 α _a _a_1 ↔ _a ≤ _a_2 ∧ _a_2 ≤ _a_1,
a x : α
⊢ (∃ (a_1 : α), a ≤ a_1 ∧ some a_1 = ↑x) ↔ a ≤ x
```

Reply:
Oh sorry, the last piece you need is
```lean
@[simps]
def coe_embedding : α ↪ with_top α :=
{ to_fun := coe, ..embedding.some}
```
Which simps to something better than `embedding.some`.